pkg/seedadmission/extension_crds_test.go

```diff
@@ -0,0 +1,571 @@
+						obj.SetKind(obj.GetKind() + "List")
+
+						c.EXPECT().List(ctx, obj, client.InNamespace(request.Namespace)).DoAndReturn(func(_ context.Context, o runtime.Object, _ ...client.ListOption) error {
+							prepareObjectWithLabelsAnnotations(o, resource, deletionProtectedLabels, deletionConfirmedAnnotations)
+							return nil
+						})
+
+						testDeletionConfirmed(ctx, c, logger, request, resource)
+					}
+				})
+			})
+		})
+	})
+})
+
+func resourceToId(resource metav1.GroupVersionResource) string {
+	return fmt.Sprintf("%s/%s/%s", resource.Group, resource.Version, resource.Resource)
+}
+
+type unstructuredInterface interface {
+	SetAPIVersion(string)
+	SetKind(string)
+}
+
+func prepareRequestAndObjectWithResource(request *admissionv1beta1.AdmissionRequest, obj unstructuredInterface, resource metav1.GroupVersionResource) {
+	request.Kind.Group = resource.Group
+	request.Kind.Version = resource.Version
+	obj.SetAPIVersion(request.Kind.Group + "/" + request.Kind.Version)
+	obj.SetKind(request.Kind.Kind)
+}
+
+func prepareObjectWithLabelsAnnotations(obj runtime.Object, resource metav1.GroupVersionResource, labels, annotations map[string]string) {
+	switch obj.(type) {
+	case *unstructured.Unstructured:
+		obj.(*unstructured.Unstructured).SetAPIVersion(fmt.Sprintf("%s/%s", resource.Group, resource.Version))
+		obj.(*unstructured.Unstructured).SetKind(resource.Resource)
+		obj.(*unstructured.Unstructured).SetLabels(labels)
+		obj.(*unstructured.Unstructured).SetAnnotations(annotations)
+	case *unstructured.UnstructuredList:
+		o := &unstructured.Unstructured{}
+		o.SetAPIVersion(fmt.Sprintf("%s/%s", resource.Group, resource.Version))
+		o.SetKind(resource.Resource)
+		o.SetLabels(labels)
+		o.SetAnnotations(annotations)
+		obj.(*unstructured.UnstructuredList).Items = []unstructured.Unstructured{*o}
+	}
+}
+
+func getObjectJSONWithLabelsAnnotations(obj *unstructured.Unstructured, resource metav1.GroupVersionResource, labels, annotations map[string]string) []byte {
+	prepareObjectWithLabelsAnnotations(obj, resource, labels, annotations)
+
+	objJSON, err := json.Marshal(obj)
+	Expect(err).NotTo(HaveOccurred())
+
+	return objJSON
+}
+
+func testDeletionUnconfirmed(
+	ctx context.Context,
+	c *mockclient.MockClient,
+	logger *logrus.Logger,
+	request *admissionv1beta1.AdmissionRequest,
+	resource metav1.GroupVersionResource,
+) {
+	request.Resource = resource
+	err := ValidateExtensionDeletion(ctx, c, logger, request)
+	Expect(err).To(HaveOccurred(), resourceToId(resource))
+	Expect(err.Error()).To(ContainSubstring("annotation to delete"), resourceToId(resource))
+}
+
+func testDeletionConfirmed(
+	ctx context.Context,
+	c *mockclient.MockClient,
+	logger *logrus.Logger,
+	request *admissionv1beta1.AdmissionRequest,
+	resource metav1.GroupVersionResource,
+) {
+	request.Resource = resource
+	Expect(ValidateExtensionDeletion(ctx, c, logger, request)).To(BeNil(), resourceToId(resource))
```

Comment:
```suggestion
	Expect(ValidateExtensionDeletion(ctx, c, logger, request)).To(Succeed(), resourceToId(resource))
```

Reply:
When do we use `BeNil()` and when do we use `Succeed()`?

Reply:
No strong opinion here, but it seems like it is good practice to use `HaveOccurred()` and `Succeed()` for asserting (non-)presence of `error`s instead of `BeNil()` for the sake of making the test a little bit easier to understand. But as said, really no strong opinion here :)